contracts/staking-vault/StakingVault.sol

```diff
@@ -0,0 +1,284 @@
+      uint256 tokenID = convertToID(tokensStaked[i], tokenBacked);
+      previousBalance = balanceStakedPerID[tokenID];
+      balanceStakedPerID[tokenID] = previousBalance - amountsDrawn[i];
+      sTokenPrice[tokenID] = (sTokenPrice[tokenID] * (previousBalance - balanceStakedPerID[tokenID])) / previousBalance;
+      unchecked {
+        ++i;
+      }
+    }
+  }
+
+  function getBackingAmount(address token) external view returns (uint256) {
+    address[] memory tokensStaked = stakingTokens;
+    uint256 value;
+    for (uint256 i; i < tokensStaked.length; ) {
+      value += priceFeed.queryReturn(tokensStaked[i], valuationToken, balanceStakedPerID[convertToID(token, tokensStaked[i])]);
+    }
+    return value;
+  }
+
+  function getBackingAmountInNativeTokens(address token) external view returns (uint256[] memory balances) {
+    address[] memory tokensStaked = stakingTokens;
+    balances = new uint256[](tokensStaked.length);
+    for (uint256 i; i < tokensStaked.length; ) {
+      balances[i] = balanceStakedPerID[convertToID(token, tokensStaked[i])];
+      unchecked {
+        ++i;
+      }
+    }
+  }
+
+  function convertToID(address depositToken, address tokenToBack) public pure returns (uint256) {
+    return uint256(keccak256(abi.encode(depositToken, tokenToBack)));
+  }
+
+  function _amountToMint(uint256 tokenID, uint256 amount) internal view returns (uint256) {
+    uint8 decimals = tokenDecimalCache[identifierToTokens[tokenID].depositToken];
+
+    return ((amount * 1e18) / sTokenPrice[tokenID]) * 10**(18 - decimals);
+  }
+
+  function _amountToSend(uint256 tokenID, uint256 amountBurnt) internal view returns (uint256) {
+    uint8 decimals = IERC20Metadata(identifierToTokens[tokenID].depositToken).decimals();
+
+    return ((sTokenPrice[tokenID] * amountBurnt) / 1e18) / 10**(18 - decimals);
+  }
+
+  function addRewards(address tokenBacked, uint256 rewardAmount) external {
+    _addRewards(rewardAmount, tokenBacked);
+    IERC20(rewardToken).safeTransferFrom(msg.sender, address(this), rewardAmount);
+
+    emit AddRewards(tokenBacked, rewardAmount);
+  }
+
+  function accumulateRewards(address tokenBacked, uint256 rewardAmount) external {
+    undistributedRewards[tokenBacked] += rewardAmount;
+    IERC20(rewardToken).safeTransferFrom(msg.sender, address(this), rewardAmount);
+
+    emit AccumulateRewards(tokenBacked, rewardAmount);
+  }
+
+  function distributeRewards(address tokenBacked) external {
+    uint256 rewards = undistributedRewards[tokenBacked];
+    _addRewards(rewards, tokenBacked);
+    undistributedRewards[tokenBacked] = 0;
+
+    emit DistributeRewards(tokenBacked, rewards);
+  }
+
+  function _addRewards(uint256 rewardAmount, address tokenBacked) internal {
+    address[] memory tokensStaked = stakingTokens;
+    uint256 totalValue;
+    uint256[] memory values = new uint256[](tokensStaked.length);
+    uint256 tokenID;
+    for (uint256 i; i < tokensStaked.length; ) {
```

Comment:
I think there must be a way to join those two loops into one

Reply:
I tried to but to calculate the splits it needs a total value which requires its own loop.